### PR TITLE
Add concurrent downloading

### PR DESCRIPTION
**Pull request by csnewman**
Download speed seems to be limited per request, not per IP or user.
This enables much faster downloading.

### CODE REVIEW

File: run.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 import string
 import sys
 from bs4 import BeautifulSoup
+import concurrent.futures
```
Comment (ed-cooper) [inline run.py]: Please place this at the top so that it is sorted like the rest of the imports

File: run.py [hunk 2]
```diff
@@ -61,72 +62,90 @@
...
-                                                       features="html.parser")
-        podcast_src = "https://video.manchester.ac.uk" + \
-                      getVideoServicePodcastPageSoup.find("video", id="video").source["src"]
 
-        # Get podcast
-        get_video_service_podcast = session.get(podcast_src, stream=True)
+with concurrent.futures.ThreadPoolExecutor(max_workers=settings.concurrent_downloads) as executor:
+    futures = []
+
+    # Status code valid, parse HTML
```
Comment (ed-cooper) [inline run.py]: The comment stating that the status code is valid should go after line 64, as from that point on we know it is valid.
Parsing HTML is it's own separate comment and should therefore remain here

File: run.py [hunk 3]
```diff
@@ -61,72 +62,90 @@
...
+    get_video_service_base_soup = BeautifulSoup(get_video_service_base.content, features="html.parser")
+    first = True
```
Comment (ed-cooper) [inline run.py]: lmao I meant to remove this, I hope you had loads of fun trying to figure out what it was for

**Review by ed-cooper — CHANGES_REQUESTED** [review]
You have modified `settings-template.py`.

Please update `README.md` to match the changes

File: run.py [hunk 4]
```diff
@@ -61,72 +62,90 @@
...
-first = True
-for course_li in get_video_service_base_soup.find("nav", {"id": "sidebar-nav"}).ul.contents[3].find_all("li", {
-        "class": "series"}):
-    # For each course
-    print("-" * (21 + len(course_li.a.string)))
-    print("Getting podcasts for", course_li.a.string)
-    print("-" * (21 + len(course_li.a.string)))
-    get_video_service_course = session.get("https://video.manchester.ac.uk" + course_li.a["href"])
+
+def download_podcast(name, podcast_link, download_path):
```
Comment (ed-cooper) [inline run.py:70]: Please add a comment explaining what this method does

**Comment by ed-cooper** [comment]
Also, did you install any modules for the concurrent.futures?

If so, please run
```bash
pip freeze > requirements.txt
```

You'll also need to pull the latest changes from the repo first if this is the case

File: run.py (hunk 3 above)
Comment (csnewman) [reply run.py]: The height of software development!

**Comment by csnewman** [comment]
> Also, did you install any modules for the concurrent.futures?
> 
> If so, please run
> 
> ```shell
> pip freeze > requirements.txt
> ```
> 
> You'll also need to pull the latest changes from the repo first if this is the case

```pip freeze``` is considered bad practice as it exposes dependencies of your dependencies. It also lists all packages installed in your virtual environment. I have instead used:
```pipreqs . --force```

**Comment by ed-cooper** [comment]
That's fine except for flake8.
Obviously this is required on the Travis CI and can be added to the `.travis.yml` file, but I also think it's useful to have it in the repo for local testing?

**Comment by ed-cooper** [comment]
What command did you use to include flake8?
Or did you just add it manually to the file?

**Comment by csnewman** [comment]
Ready for second round of review.

**Review by ed-cooper — APPROVED** [review]

**Comment by ed-cooper** [comment]
I think set the default concurrent streams to 4 for people on WiFi (both in the template and README)

File: README.md [hunk 5]
```diff
@@ -14,6 +14,7 @@ Then, open it in your editor of your choice, and set the following 3 variables:
 username = "Your username"        # The username you would usually use for My Manchester
 password = "Your password"        # The accompanying password
 base_dir = "~/Documents/Lectures" # Where to download files to
+concurrent_downloads = 8          # How many podcasts to download concurrently (Increases speed)
```
Comment (ed-cooper) [inline README.md]: Use simultaneously instead of concurrently for the description

**Review by ed-cooper — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (csnewman) [reply README.md]: Concurrent is the correct term here, as the downloads are overlapping over a period of time, whereas simultaneously refers to a occurring at the same instant.

**Comment by csnewman** [comment]
> I think set the default concurrent streams to 4 for people on WiFi (both in the template and README)

Default lowered.

**Comment by ed-cooper** [comment]
Yes, but most people don't know what concurrent means, whereas simultaneous
makes more sense for normal people

**Comment by csnewman** [comment]
> Yes, but most people don't know what concurrent means, whereas simultaneous makes more sense for normal people

In that case, they will not mess with the setting. We can link to a dictionary or StackOverflow post explaining the meaning if it becomes an issue.

**Comment by ed-cooper** [comment]
I held a democratic poll at brunch and the results were:
Simultaneous - 5
Concurrent - 0

**Comment by csnewman** [comment]
> I held a democratic poll at brunch and the results were: Simultaneous - 5 Concurrent - 0

I sadly do not accept democracy and ran a second referendum on this issue:
Simultaneous = -5
Concurrent = +10000

**Comment by ed-cooper** [comment]
Fine, I concede.

Now add a link to the definition in the readme.

**Review by ed-cooper — DISMISSED** [review]

**Comment by csnewman** [comment]
Where would you suggest the link being placed. As it is not possible to have a link inside a code fragment

**Comment by ed-cooper** [comment]
How about after the code fragment?

I was thinking having a whole paragraph dedicated to explaining the meaning of concurrency and why they are so much more than simultaneous downloads.

**Comment by csnewman** [comment]
A peer reviewed paper might be the only solution to that problem. 

However in the meantime, the wording has been changed.

**Review by ed-cooper — APPROVED** [review]
👍

**Pull request merged**